watcher/handlers/metalogger.go

```diff
@@ -122,18 +144,145 @@ func (l *MetaLogger) OnAdd(obj interface{}) error {
 			"__kf_workspace__": mlpbStringValue(workspace),
 		},
 	}
-	request := storepb.PutArtifactsRequest{
-		Artifacts: []*mlpb.Artifact{&artifact},
+	request := storepb.PutExecutionsRequest{
+		Executions: []*mlpb.Execution{execution},
 	}
-	resp, err := l.kfmdClient.PutArtifacts(context.Background(), &request)
+	resp, err := l.kfmdClient.PutExecutions(context.Background(), &request)
 	if err != nil {
-		klog.Errorf("failed to log metadata for %s: err = %s, request = %v, resp = %v", l.MetadataArtifactType(), err, request, resp)
+		klog.Errorf("failed to log metadata for %s: err = %s, request = %v, resp = %v", k8smlmd.GVKToType(l.resource), err, request, resp)
 		return err
 	}
-	klog.Infof("Handled addEvent for %s.\n", l.MetadataArtifactType())
+
+	if len(resp.ExecutionIds) == 0 {
+		err := fmt.Errorf("No execution id returned from MLMD")
+		log.Error(err, "Newly created execution is missing an id")
+		return err
+	}
+
+	for  _, annotation := range []string{v1alpha1.InputAnnotation, v1alpha1.OutputAnnotation} {
+		ids, err := l.logArtifactAnnotation(object, annotation)
+
+		if err != nil {
+			log.Error(err, "Could not log artifact annotation", "annotation", annotation)
+			continue
+		}
+
+		// Create events connecting the execution and artifacts
+		for _, aId := range ids {
+			// TODO(jlewi): What should we set for the path if anythong?
+			e := &mlpb.Event{
+				ArtifactId: proto.Int64(aId),
+				ExecutionId: proto.Int64(resp.ExecutionIds[0]),
+			}
+
+			if annotation == v1alpha1.InputAnnotation {
+				e.Type = mlpb.Event_INPUT.Enum()
+			} else {
+				e.Type = mlpb.Event_OUTPUT.Enum()
+			}
+
+			eRequest := &storepb.PutEventsRequest{
+				Events:               []*mlpb.Event{e},
+			}
+			_, err := l.kfmdClient.PutEvents(context.Background(), eRequest)
+
+			if err != nil {
+				log.Error(err, "Failed to create event", "event", e)
+			}
+		}
+	}
+
+	log.Info("Handled addEvent")
 	return nil
 }
 
+// logArtifactAnnotation looks for the specified annotation and if it is present tries to create an
+// artifact storing that value.
+func (l *MetaLogger) logArtifactAnnotation(object metav1.Object, annotation string) ([]int64, error) {
+	log := l.log.WithValues("name", object.GetName(), "uid", object.GetUID())
+
+	input, ok := object.GetAnnotations()[annotation]
+
+	if !ok {
+		return []int64{}, nil
+	}
+
+	inputArtifact, err := toUnstructured(input)
+
+	if err != nil {
+		return []int64{}, err
+	}
+
+	gvk := inputArtifact.GetObjectKind().GroupVersionKind()
```

Comment:
This requires us to have GVK for the input. this makes us define more crds and build common wrapper on the inputs like `dataset`, `persist_model` etc. 

Do we want to support primitive type like a  string ?

Reply:
It doesn't require CRDs. It requires that metadata schemas adhere to the Kubernetes resource model (https://docs.google.com/document/d/1RmHXdLhNbyOWPW_AtnnowaRfGejw-qlKQIuLKQWlwzs/edit#) which really just requires that they have metadata.

That seems like a good thing because that enables introspection just like we are doing here.

An example of a dataset is
https://github.com/kubeflow/metadata/blob/96285b257d999715a88befc659d722c895d7bbc5/watcher/examples/dataset.yaml

So we just need to include the fields "apiVersion and kind". That doesn't seem overly burdensome and offers a lot of nice properties.

Reply:
I misunderstand that. I thought `dataset` was crd but actually a definition used by manifest only. then it sounds good to me